File: blue-sdk-android/blue-sdk/src/main/kotlin/com/blue/sdk/manager/AudioManager.kt
// AudioManager.kt
// BlueSDK - 音频与系统设置管理器（FR25~FR31）

package com.blue.sdk.manager

import com.blue.sdk.enums.SoundType
import com.blue.sdk.enums.TimeFormat
import com.blue.sdk.enums.VolumeLevel
import com.blue.sdk.error.BlueError
import com.blue.sdk.internal.CommandQueue
import com.blue.sdk.transport.CommandCode
import com.blue.sdk.transport.DPIDConstants
import com.blue.sdk.transport.FrameBuilder

internal class AudioManager(private val commandQueue: CommandQueue) {

    /**
     * 设置音量等级
     * 使用 DPID 0x6E（ALERT_DURATION），type=0x04
     * 帧格式：78 04 00 01 XX（01低/02中/03高）
     * 协议示例：55 AA 00 06 00 05 78 04 00 01 01 88
     */
    fun setVolume(level: VolumeLevel, completion: (Result<Unit>) -> Unit) {
        sendCommand(byteArrayOf(0x78.toByte(), 0x04, 0x00, 0x01, level.protocolValue), completion)
    }

    /**
     * 设置铃声类型
     * 使用 DPID 0x6D（TYPE_OF_SOUND）
     * 帧格式：6D 04 00 01 XX（00=静音/01=A/02=B/03=C）
     */
    fun setSoundType(type: SoundType, completion: (Result<Unit>) -> Unit) {
        sendCommand(byteArrayOf(DPIDConstants.TYPE_OF_SOUND, 0x04, 0x00, 0x01, type.protocolValue), completion)
    }

    /** 设置静音（通过铃声类型=0x00实现，取消静音恢复为类型A）*/
    fun setSilence(enabled: Boolean, completion: (Result<Unit>) -> Unit) {
        if (enabled) setSoundType(SoundType.MUTE, completion)
        else setSoundType(SoundType.TYPE_A, completion)
    }

    /**
     * 设置提醒持续时间（分钟）
     * 使用 DPID 0x70（EMPTY_ALL_ALARMS），type=0x02
     * 帧格式：6E 02 00 04 00 00 00 XX
     * 协议示例：55 AA 00 06 00 08 6E 02 00 04 00 00 00 05 86（2分钟，值=5）
     */
    /**
     * 设置提醒持续时长（1~5分钟）
     * Set alert duration (1~5 minutes)
     */
    fun setAlertDuration(minutes: Int, completion: (Result<Unit>) -> Unit) {
        if (minutes < 1 || minutes > 5) {
            completion(Result.failure(BlueError.InvalidParameter))
            return
        }
        sendCommand(byteArrayOf(DPIDConstants.ALERT_DURATION, 0x02, 0x00, 0x04,
            0x00, 0x00, 0x00, minutes.toByte()), completion)
    }

    fun setTimeFormat(format: TimeFormat, completion: (Result<Unit>) -> Unit) {
        sendCommand(byteArrayOf(DPIDConstants.TIME_FORMAT, 0x04, 0x00, 0x01, format.protocolValue), completion)
    }

    private fun sendCommand(data: ByteArray, completion: (Result<Unit>) -> Unit) {
        val frame = FrameBuilder.build(CommandCode.SEND_COMMAND, data)
        commandQueue.enqueue(CommandCode.SEND_COMMAND, frame) { result ->
            result.fold(
                onSuccess = { completion(Result.success(Unit)) },
                onFailure = { completion(Result.failure(it as BlueError)) }
            )
        }
    }

    companion object {
        fun parseSoundType(data: ByteArray): SoundType? =
            if (data.size >= 5) SoundType.fromByte(data[4]) else null

        fun parseTimeFormat(data: ByteArray): TimeFormat? =
            if (data.size >= 5) TimeFormat.values().find { it.protocolValue == data[4] } else null
    }
}

File: blue-sdk-ios/BlueSDK/BlueSDK/Classes/Manager/AudioManager.swift
// AudioManager.swift
// BlueSDK - LX-PD02 智能药盒蓝牙通信 SDK
//
// 音频与系统设置管理器（FR25~FR31）
// 协议 DPID 对应关系：
//   0x6D - 声音类型（0-静音 1-声音A 2-声音B 3-声音C）— APP 下发 & 设备上报
//   0x6E - 提醒持续时间 / 音量设置（通过 type 字段区分）
//   0x6F - 用药结果通知（APP→设备）
//   0x73 - 时制（0-12H 1-24H）
//   0x74 - 当前闹钟静音（0-关 1-开）

import Foundation

/// 音频与系统设置管理器
final class AudioManager {

    private let commandQueue: CommandQueue
    private let logger = BlueLogger.shared

    init(commandQueue: CommandQueue) {
        self.commandQueue = commandQueue
    }

    // MARK: - 声音类型设置（DPID=0x6D）

    /// 设置设备铃声类型
    /// APP 下发用 DPID=0x6D（TYPEOFSOUND），值：00=静音 01=类型A 02=类型B 03=类型C
    func setSoundType(_ type: SoundType, completion: @escaping (Result<Void, BlueError>) -> Void) {
        let data: [UInt8] = [DPIDConstants.typeOfSound, 0x04, 0x00, 0x01, type.protocolValue]
        sendCommand(data: data, completion: completion)
    }

    // MARK: - 提醒持续时间设置（DPID=0x70）

    /// 设置提醒持续时长（分钟）
    /// 设置提醒持续时长
    /// 帧格式：6E 02 00 04 00 00 00 [值]
    /// 设置提醒持续时长（1~5分钟）
    /// 协议示例：55 AA 00 06 00 08 6E 02 00 04 00 00 00 05 86
    func setAlertDuration(_ minutes: Int, completion: @escaping (Result<Void, BlueError>) -> Void) {
        guard minutes >= 1 && minutes <= 5 else {
            completion(.failure(.invalidParameter))
            return
        }
        let data: [UInt8] = [
            DPIDConstants.alertDuration,
            0x02, 0x00, 0x04,
            0x00, 0x00, 0x00,
            UInt8(minutes)
        ]
        sendCommand(data: data, completion: completion)
    }

    // MARK: - 音量设置

    /// 设置设备提醒音量
    /// 协议示例：55 AA 00 06 00 05 78 04 00 01 01 88（低音量）
    /// DPID=0x78, type=0x04, len=0x01, value: 01=低 02=中 03=高
    func setVolume(_ level: VolumeLevel, completion: @escaping (Result<Void, BlueError>) -> Void) {
        let data: [UInt8] = [0x78, 0x04, 0x00, 0x01, level.protocolValue]
        sendCommand(data: data, completion: completion)
    }

    // MARK: - 静音设置

    /// 设置静音（实际通过铃声类型=0x00实现）
    /// enabled=true 时设置铃声类型为静音，enabled=false 时恢复为类型A
    func setSilence(_ enabled: Bool, completion: @escaping (Result<Void, BlueError>) -> Void) {
        if enabled {
            setSoundType(.mute, completion: completion)
        } else {
            setSoundType(.typeA, completion: completion)
        }
    }

    // MARK: - 时间格式设置（DPID=0x73）

    func setTimeFormat(_ format: TimeFormat, completion: @escaping (Result<Void, BlueError>) -> Void) {
        let data: [UInt8] = [DPIDConstants.timeFormat, 0x04, 0x00, 0x01, format.protocolValue]
        sendCommand(data: data, completion: completion)
    }

    // MARK: - 解析上报帧

    /// 解析铃声类型上报（DPID=0x6D）
    /// 设备上报值：0=静音, 1=类型A, 2=类型B
    static func parseSoundType(from data: [UInt8]) -> SoundType? {
        guard data.count >= 5 else { return nil }
        let value = data[4]
        return SoundType.from(byte: value)
    }

    /// 解析时间格式上报（DPID=0x73）
    static func parseTimeFormat(from data: [UInt8]) -> TimeFormat? {
        guard data.count >= 5 else { return nil }
        return TimeFormat(rawValue: Int(data[4]))
    }

    // MARK: - 私有方法

    private func sendCommand(data: [UInt8], completion: @escaping (Result<Void, BlueError>) -> Void) {
        let frame = FrameBuilder.build(cmd: CommandCode.sendCommand, data: data)
        commandQueue.enqueue(cmd: CommandCode.sendCommand, frame: frame) { result in
            switch result {
            case .success:
                completion(.success(()))
            case .failure(let error):
                completion(.failure(error))
            }
        }
    }
}
